### ISSUE TITLE
Convert function comments into docstrings, and generally start using docstrings

### ISSUE THREAD

**Issue by audiodude** [noisebridge/MediaBridge#29]
[PEP-257](https://peps.python.org/pep-0257/) talks about docstrings for functions and methods. Although we are unlikely to ever generate documentation automatically for this code, it would still be nice to have some docstrings. I recommend adding them on a "best effort" basis.

**Comment by audiodude** [comment]
As requested in #17

**Comment by skyfenton** [comment]
Is the idea to document functions and things with like google's doc template? Not that we should rely on extensions for it, but [this one](https://marketplace.visualstudio.com/items?itemName=njpwerner.autodocstring) makes it really easy to generate docs. Idk if we want to go down the more self-documenting route, though via type hints and mostly single line docstrings.

**Comment by jhanley634** [comment]
Adding a one-sentence docstring to a function is an excellent idea, though some functions are simple enough that a name of `def get_foo():` is self-documenting, without need of a """Get the foo.""" sentence.

Adding additional sentences is sometimes warranted. In the same vein, adding tabular or prose descriptions of parameters is sometimes warranted. Perhaps in the format suggested by google's doc template, something which predates modern python type annotations by many years.

Adding type annotations to a signature is always helpful. And we can use `$ mypy --strict .` to lint it.

A significant downside of /** javadoc */ comments, /// csharp doc comments, and """python docstrings""" is that they bit rot. That is, they start out accurate. Then someone changes a signature, and doesn't update the docstring, and a reviewer doesn't catch that. This happens with parameter renames, inserted param, deleted param. The signature is what matters. Prefer to document a function in the signature if feasible, by using helpful types and identifier names. Take on the maintenance burden of writing additional documentation if greater detail is still needed.

**Comment by skyfenton** [comment]
The question then is, should we enforce either type hints or docstrings with ruff? I lean toward no and we should just request descriptive docstrings in PRs if a function feels like it needs one, but to avoid infinite nits, maybe it's worth enforcing at least a one line docstring for each function and optionally type hints. Unless we do want to enforce full docstrings where we describe each input/output/return with types, and just ignore type hints unless we want them.

 It'd be a lot simpler of a decision if type hints were better than they are now, but they're more cumbersome in a lot of cases imo. Though to be fair, I don't think our cases are especially unwieldly and maybe being stricter in our code enforcement earlier in dev until we run into an issue will avoid a lot of refactoring down the line if we want to be stricter later.

**Comment by audiodude** [comment]
Is ruff giving you warnings for docstrings? We can definitely turn that off if it's the case.

**Comment by audiodude** [comment]
I agree that if we're going to use typing at all, we should start doing it now.

**Comment by jhanley634** [comment]
tl;dr: Annotation good, mandatory docstring bad.


@skyfenton , I can weigh in on this. As usual, the reference repo I cite as evidence for what's workable for a similarly diverse team is: https://github.com/jhanley634/dojo-blackboard

I concede that there _is_ a bit of a learning curve for type hinting, and I apologize for it. Python is not java; I feel most of the benefit comes from human readable documentation, and just a little bit comes from {pyright, mypy} type safety proofs. They _can_ be helpful, but they're not the primary motivation. Keeping type documentation out of human-readable docstrings is good. I like documentation I can believe, such as mypy verified signatures, or python -m doctest *.py. Sometimes "comments lie!", and I _love_ seeing ">>>" in a docstring since I know I can believe it, I know the machine has verified it.

A simple "[pyright](https://pypi.org/project/pyright) ." is fast and gives reasonably good diagnostic messages. Having passed that, maybe follow it with some form of "mypy --strict", which will _insist_ on seeing params and result types called out.

Here's my annotation advice:

- It would be good for "please annotate!" to be a social norm.
- It's appropriate to ask for annotations in a PR, but they shouldn't be mandatory.
- It will take one or more months for the team to become familiar with the details. Practice makes perfect. Annotated python is a related but different language from what you've been writing in.
- I do not encourage automated git .pre-commit-config.yaml hooks or GitHub Actions which demand annotation.
- I do encourage motivated team members to make occasional cleanup passes, adding annotations, to remove speedbumps and make it easier for others to adopt routine annotating and linting.

----

I have seen good results with a bunch teams adopting this advice: "black . && ruff check" should lint clean. It is easy to adopt, and ruff's defaults won't get in your way. Anything it calls out, you should just tidy up so it lints clean. If ruff didn't like it, then it's a good bet a human reviewer wouldn't like it, so deal with it early in order to have a productive PR cycle. And uniform rules are good -- there's no pointless arguments about details that don't matter.

In this project there's no `black`. Sometimes I add the occasional extra NEWLINE between function definitions, so `git status` lets me copy-n-paste the newly dirtied filename, confident that things will soon be automatically tidied up. I was a little surprised to notice that `ruff check` didn't call me on the ugly dozen blank lines I had introduced; it's simply not a use case I had ever explored. Now, `ruff` has many checks that can be optionally enabled, it's very flexible. It turns out that configuring "E3", and "ruff check --preview", will signal E303 Too many blank lines. But I feel --preview rules are not appropriate for a project-wide standard, so I wouldn't go there, preferring `black` instead. It also tidies up things like "W291 Trailing whitespace" and "W293 Blank line contains whitespace", so I literally never see such issues in projects I work on. The git commit hook already dealt with it by the time the code hit the repository.

Docstrings are not always necessary, and linters should not auto-flag if one is missing. Humans can evaluate the need better than AST parsing tools can. Some simple getters are self-documenting, just from the signature, especially if it is annotated. The rule is: Signature + optional docstring should be enough for a maintainer to understand the [contract](https://en.wikipedia.org/wiki/Design_by_contract) and write a sensible unit test.

I like `ruff` defaults. I do not recommend that you follow the learning adventure I embarked on with [this config](https://github.com/jhanley634/dojo-blackboard/blob/ed6b8c5a/pyproject.toml#L42-L66). I experimented with the "D" doc lint rules. One by one I disabled nearly all of them as I encountered code they flagged where I disagreed with the linter's advice.

BTW, rather than one .ini or .toml config file for each tool, you might consider putting all their configs into pyproject.toml.

Summary: I encourage gradual yet relentless adoption of annotation. It's good for you, and it enables beartype, numba, typer, along with other parts of the ecosystem.

**Comment by audiodude** [comment]
Fixed by #37
